src/devices/bus/epson_qx/multifont.cpp

```diff
@@ -0,0 +1,258 @@
+	ROM_REGION( 0x4000, "font6", 0 )
+	ROM_LOAD( "hn43128.1c", 0x0000, 0x4000, NO_DUMP)
+	ROM_REGION( 0x4000, "font7", 0 )
+	ROM_LOAD( "hn43128.2c", 0x0000, 0x4000, NO_DUMP)
+ROM_END
+
+INPUT_PORTS_START(multifont)
+	PORT_START("IOBASE")
+	PORT_CONFNAME(0x06, 0x04, "IO Base Address Selection")
+	PORT_CONFSETTING(0x00, "&F8")
+	PORT_CONFSETTING(0x02, "&FA")
+	PORT_CONFSETTING(0x04, "&FC")
+	PORT_CONFSETTING(0x06, "&FE")
+INPUT_PORTS_END
+
+//-------------------------------------------------
+//  multifont_device - constructor
+//-------------------------------------------------
+multifont_device::multifont_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	device_t(mconfig, EPSON_QX_OPTION_MULTIFONT, tag, owner, clock),
+	device_option_expansion_interface(mconfig, *this),
+	m_mcu(*this, "mcu"),
+	m_i8243(*this, "i8243"),
+	m_fonts(*this, "font%u",0),
+	m_iobase(*this, "IOBASE")
+{
+}
+
+//-------------------------------------------------
+//  device_input_ports - device-specific ports
+//-------------------------------------------------
+ioport_constructor multifont_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME(multifont);
+}
+
+//-------------------------------------------------
+//  device_rom_region - device-specific roms
+//-------------------------------------------------
+const tiny_rom_entry *multifont_device::device_rom_region() const
+{
+	return ROM_NAME(multifont);
+}
+
+//-------------------------------------------------
+//  device_add_mconfig - device-specific config
+//-------------------------------------------------
+void multifont_device::device_add_mconfig(machine_config &config)
+{
+	I8039(config, m_mcu, 9_MHz_XTAL);
+	m_mcu->set_addrmap(AS_PROGRAM, &multifont_device::rom_map);
+	m_mcu->prog_out_cb().set(m_i8243, FUNC(i8243_device::prog_w));
+	m_mcu->p1_out_cb().set(FUNC(multifont_device::p1_w));
+	m_mcu->p2_out_cb().set(FUNC(multifont_device::p2_w));
+	m_mcu->bus_out_cb().set(FUNC(multifont_device::write_bus));
+	m_mcu->bus_in_cb().set(FUNC(multifont_device::read_bus));
+	m_mcu->t1_in_cb().set(FUNC(multifont_device::t1_r));
+	m_mcu->t0_in_cb().set(FUNC(multifont_device::t0_r));
+	I8243(config, m_i8243);
+	m_i8243->p4_out_cb().set(FUNC(multifont_device::rom_addr_w<0>));
+	m_i8243->p5_out_cb().set(FUNC(multifont_device::rom_addr_w<4>));
+	m_i8243->p6_out_cb().set(FUNC(multifont_device::rom_addr_w<8>));
+	m_i8243->p7_out_cb().set(FUNC(multifont_device::rom_select_w));
+}
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+void multifont_device::device_start()
+{
+}
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+void multifont_device::device_reset()
+{
+	m_data_out = 0x00;
+	m_data_in = 0x00;
+	m_p2 = 0x00;
+	m_status = 0x00;
```

Comment:
You shouldn’t change values like this on reset.  If the 8039 port outputs change on reset, it should push out the new values.  IIRC these values aren’t even the correct reset values – at least the P2 output goes high-impedance high on reset.

Reply:
So m_status at least should be set on reset, its values are stored in flipflops that are reset via the reset signal. The others I will move to device_start for initialization. m_data_in/out are two 74ls373, but aren't attached to the reset line.

So when looking at the mcs48 code, it seems like its performing a bus_w during reset, which to me looks wrong. I mean the 8039 doesn't actually write anything to the bus during reset, it just has the pins in high impedance so that if anyone else reads from it they will get the value 0xff. Doing a bus_w actually triggers my callback which sets the IBF flag on the multifont card during reset which is clearly not correct, since there is no actual data waiting to be read. Same thing for p1 and p2. Am i missing something about how this is supposed to work?

Reply:
The thing with the bus pins is that they can be used as GPIO (using OUTL/ANL/ORL and INS) as well as being read or written synchronously (using MOVX).  The OUTL/ANL/ORL and INS instruction do generate /RD and /WR strobes, but they aren’t generally used when it’s used for GPIO like this.

Which instructions is is the card using for bus I/O?  Is it using the /RD and /WR strobes?

Reply:
The `bus_w` callback is supposed to be used when the bus is used for GPIO.  You’re supposed to use I/O space handlers when it’s using the MOVX instructions.  After a cursory glance at the schematic, I suspect that is how it’s doing bus I/O.

That aside, the reason it calls `bus_w` on reset is because if you’d used OUTL or ANL to drive bus pins low using the bus output latch, the lines are released on reset.  This is visible to external devices.  It isn’t catering for the unusual use case of using the bus output latch and also monitoring the /WR line.

Reply:
Ok I see at least what the issue is, i just assumed that the _w implied that an actual write was being done. While not really ideal I ended up using a flag in my code to prevent updating the status and ouput latch during the reset condition.